pyschema/core.py

```diff
@@ -85,24 +85,29 @@ def add_record(self, schema, _bump_stack_level=False):
             Can be used as a class decorator
         """
         full_name = get_full_name(schema)
-        self._force_add(full_name, schema, _bump_stack_level)
-        if '.' in full_name and schema.__name__ not in self._schema_map:
+        has_namespace = '.' in full_name
+        self._force_add(full_name, schema, _bump_stack_level, _raise_on_existing=has_namespace)
+        if has_namespace and schema.__name__ not in self._schema_map:
             self._force_add(schema.__name__, schema, _bump_stack_level)
         return schema
 
-    def _force_add(self, used_name, schema, _bump_stack_level=False):
+    def _force_add(self, used_name, schema, _bump_stack_level=False, _raise_on_existing=False):
         existing = self._schema_map.get(used_name, None)
-        if existing:
-            full_name = get_full_name(schema)
-            explanation = "(actually {0})".format() if full_name != used_name else ""
-
-            warnings.warn(
-                "{used_name}{explanation}: old definition in {prev_module} replaced by definition in {new_module}"
-                .format(used_name=used_name,
-                        explanation=explanation,
-                        prev_module=existing.__module__,
-                        new_module=schema.__module__),
-                stacklevel=4 if _bump_stack_level else 3)
+        if existing and existing != schema:
+            if _raise_on_existing:
+                raise ValueError('Attempted to add a new schema replacing an existing one. '
```

Comment:
I think the exception should be raised when registing the second schema. Due to the auto registration that happens on import however and I agree that that might not be the best time to raise.

I like your idea of replacing the schema with something that will raise on instantiation. The current issue is that the exception presented to the user was hard to interpret. Having a specific duplicate schema exception sounds like a good idea.